lib/krb5/build_auth.c

```diff
@@ -73,10 +74,93 @@ make_etypelist(krb5_context context,
 	krb5_abortx(context, "internal error in ASN.1 encoder");
     free_EtypeList(&etypes);
 
-    ret = _krb5_add_1auth_data(context,
-                               KRB5_AUTHDATA_GSS_API_ETYPE_NEGOTIATION, &e, 0,
-                               auth_data);
+    ade.ad_type = KRB5_AUTHDATA_GSS_API_ETYPE_NEGOTIATION;
+    ade.ad_data = e;
+
+    ret = add_AuthorizationData(auth_data, &ade);
+
     krb5_data_free(&e);
+
+    return ret;
+}
+
+/* HACK: temp */
+static void
+encode_le_uint32(uint32_t n, uint8_t *p)
+{
+    p[0] = (n >> 0 ) & 0xFF;
+    p[1] = (n >> 8 ) & 0xFF;
+    p[2] = (n >> 16) & 0xFF;
+    p[3] = (n >> 24) & 0xFF;
+}
+
+/* Flag in KRB5_AUTHDATA_AP_OPTIONS */
+#define KERB_AP_OPTIONS_CBT 0x00004000
+
+static krb5_error_code
+add_ap_options(krb5_context context,
+	       krb5_authdata *auth_data)
+{
+    krb5_error_code ret;
+    AuthorizationDataElement ade;
+    krb5_boolean require_cb;
+    uint8_t ap_options[4];
+
+    require_cb = krb5_config_get_bool_default(context, NULL, FALSE,
```

Comment:
I'm not sure whether a global configuration directive is the right way to request this. It should likely at least be per application service name (first part of SPN), and/or the initiator setting `GSS_C_CB_CONFIRM_FLAG`. @nicowilliams, you are the expert on this, what say you?

Reply:
Yeah, we need a way to set it in the API, but on one hand it's not clear how the API would look like, and on the other hand we also need global config so that old applications could still use it. So I opted to start with the global config which is simpler.

See the (long) discussion on krbdev (we also discussed it on the krbdev call with Nico):
http://mailman.mit.edu/pipermail/krbdev/2020-March/013241.html

In short, by having it in global krb5.conf we can protect any existing app which operates not over TLS (say the server requires signing when not over TLS, like Windows), as it prevent a mitm to take their token and use it over TLS without signing and without channel-bindings.  Since with this flag in the token, the server will not allow all zeros when over TLS (that is, when acceptor passed bindings).